File: 2021/June/Day 12/complex_number_multiplication.py
# source : https://leetcode.com/problems/complex-number-multiplication/submissions/
class Solution(object):
    def complexNumberMultiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        r1,c1 = self.splitc(num1);
        r2,c2 = self.splitc(num2);
        r3 = r1*r2 - (c1*c2);
        c3 = r1*c2+r2*c1;
        out = str(r3) + "+"+str(c3)+"i"
        return out
    def splitc(self,num1):
        plus1=0;
        r1=""
        c1=""

        for i in num1:
            if(i!='+' and not plus1):
                r1+=i
            elif(i=='+'):
                plus1=1;
            elif(plus1 and i!='i'):
                 c1+=i;
        return int(r1),int(c1);

            

File: 2023/May/Day 1/complexNumberMultiplication.cpp
/*

Source : https://leetcode.com/problems/complex-number-multiplication/submissions/

Logic : 
1) Extract the real and complex parts from the string, use a method for that.
2) Multiply real by real, complex by complex for the real part. THen do the remaining for complex part.
3) Return the value.

*/


class Solution {
public:
    int VERBOSE=0;
    
    vector<int> getComponents(string nums1)
    {
        int real=0;
        int complex=0;
        int f=0;
        
        string tmp="";
        for(auto&i:nums1)
        {
            if((i!='+') && (i!='i'))
            {
                tmp+=i;
            }
            else if(i=='+')
            {
                real=stoi(tmp);
                tmp="";
            }
            
            else if(i=='i')
            {
                f=1;
            }
            
            
        }
        
        if(f==1)
        {
            complex=stoi(tmp);
        }
        else
        {
            real=stoi(tmp);
        }
        if(VERBOSE)
        cout<<"Real : "<<real<<" complex : "<<complex<<endl;
        vector<int> out;
        out.insert(out.end(),real);
        out.insert(out.end(),complex);
        return out;
        
    }
    
    string complexNumberMultiply(string num1, string num2) {
        
        vector<int> v1=getComponents(num1);
        vector<int> v2=getComponents(num2);
        
        int real = v1[0]*v2[0] - (v1[1]*v2[1]);
        int complex= v1[0]*v2[1] + (v2[0]*v1[1]);
        string tmp="";
        tmp+=to_string(real)+"+"+to_string(complex)+"i";
        return tmp;
        
        
    }
};